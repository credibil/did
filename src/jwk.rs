//! # DID Key
//!
//! The `did:key` method is a DID method for static cryptographic keys. At its
//! core, it is based on expanding a cryptographic public key into a DID
//! Document.
//!
//! See:
//!
//! - <https://w3c-ccg.github.io/did-method-key>
//! - <https://w3c.github.io/did-resolution>

pub mod operator;
pub mod resolver;

/// Receiver for the `did:jwk` method.
/// TODO: Remove this. Just need the namespace, not a struct.
#[allow(clippy::module_name_repetitions)]
pub struct DidJwk;
