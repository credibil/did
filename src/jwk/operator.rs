//! # DID Key Operations
//!
//! Implements Create, Read, Update, Delete (CRUD) operations for DID Key.
//!
//! See <https://w3c-ccg.github.io/did-method-key>

use anyhow::anyhow;
use base64ct::{Base64UrlUnpadded, Encoding};
use curve25519_dalek::edwards::CompressedEdwardsY;
use serde_json::json;

use super::DidJwk;
use crate::core::Kind;
use crate::document::{CreateOptions, Document, MethodType, PublicKeyFormat, VerificationMethod};
use crate::error::Error;
use crate::{DidOperator, KeyPurpose};

impl DidJwk {
    /// Create operation for `did:jwk`.
    /// 
    /// # Errors
    /// TODO: Document errors
    pub fn create(op: &impl DidOperator, options: CreateOptions) -> crate::Result<Document> {
        let Some(verifying_key) = op.verification(KeyPurpose::VerificationMethod) else {
            return Err(Error::Other(anyhow!("no verification key")));
        };

        let serialized = serde_json::to_vec(&verifying_key)
            .map_err(|e| Error::Other(anyhow!("issue serializing key: {e}")))?;
        let encoded = Base64UrlUnpadded::encode_string(&serialized);
        let did = format!("did:jwk:{encoded}");

        // key agreement
        // <https://w3c-ccg.github.io/did-method-key/#encryption-method-creation-algorithm>
        let key_agreement = if options.enable_encryption_key_derivation {
            let key_bytes = Base64UrlUnpadded::decode_vec(&verifying_key.x)
                .map_err(|e| Error::InvalidPublicKey(format!("issue decoding key: {e}")))?;

            // derive an X25519 public encryption key from the Ed25519 key
            let edwards_y = CompressedEdwardsY::from_slice(&key_bytes).map_err(|e| {
                Error::InvalidPublicKey(format!("public key is not Edwards Y: {e}"))
            })?;
            let Some(edwards_pt) = edwards_y.decompress() else {
                return Err(Error::InvalidPublicKey(
                    "Edwards Y cannot be decompressed to point".into(),
                ));
            };
            let x25519_bytes = edwards_pt.to_montgomery().to_bytes();

            let mut jwk = verifying_key.clone();
            jwk.x = Base64UrlUnpadded::encode_string(&x25519_bytes);
            let method_type = MethodType::JsonWebKey2020;
            let key_format = PublicKeyFormat::PublicKeyJwk { public_key_jwk: jwk };

            Some(vec![Kind::Object(VerificationMethod {
                id: format!("{did}#key-1"),
                controller: did.clone(),
                type_: method_type,
                key: key_format,
                ..VerificationMethod::default()
            })])
        } else {
            None
        };

        let verif_type = &options.method_type;
        let context = Kind::Object(json!({
            "publicKeyJwk": {
                "@id": "https://w3id.org/security#publicKeyJwk",
                "@type": "@json"
            },
            verif_type.to_string(): format!("https://w3id.org/security#{verif_type}"),
        }));

        let kid = format!("{did}#key-0");

        let key_format = match options.method_type {
            MethodType::Multikey
            | MethodType::Ed25519VerificationKey2020
            | MethodType::X25519KeyAgreementKey2020 => {
                // multibase encode the public key
                PublicKeyFormat::PublicKeyMultibase {
                    public_key_multibase: verifying_key.to_multibase()?,
                }
            }
            MethodType::JsonWebKey2020 | MethodType::EcdsaSecp256k1VerificationKey2019 => {
                PublicKeyFormat::PublicKeyJwk {
                    public_key_jwk: verifying_key,
                }
            }
        };

        Ok(Document {
            context: vec![Kind::String(options.default_context), context],
            id: did.clone(),
            verification_method: Some(vec![VerificationMethod {
                id: kid.clone(),
                controller: did,
                type_: options.method_type,
                key: key_format,
                ..VerificationMethod::default()
            }]),
            authentication: Some(vec![Kind::String(kid.clone())]),
            assertion_method: Some(vec![Kind::String(kid.clone())]),
            capability_invocation: Some(vec![Kind::String(kid.clone())]),
            capability_delegation: Some(vec![Kind::String(kid)]),
            key_agreement,
            ..Document::default()
        })
    }

    #[allow(dead_code)]
    /// Read operation for `did:jwk`.
    /// 
    /// # Errors
    /// TODO: Document errors
    pub fn read(_did: &str, _: CreateOptions) -> crate::Result<Document> {
        // self.resolve(did, options)
        unimplemented!("read")
    }
}

#[cfg(test)]
mod test {
    use credibil_infosec::{Curve, KeyType, PublicKeyJwk};
    use ed25519_dalek::SigningKey;
    use rand::rngs::OsRng;

    use super::*;

    #[test]
    fn create() {
        let mut options = CreateOptions::default();
        options.enable_encryption_key_derivation = true;

        let op = Operator;
        let res = DidJwk::create(&op, options).expect("should create");

        let json = serde_json::to_string_pretty(&res).expect("should serialize");
        println!("{json}");
    }

    struct Operator;
    impl DidOperator for Operator {
        fn verification(&self, purpose: KeyPurpose) -> Option<PublicKeyJwk> {
            match purpose {
                KeyPurpose::VerificationMethod => {
                    let key = generate();

                    Some(PublicKeyJwk {
                        kty: KeyType::Okp,
                        crv: Curve::Ed25519,
                        x: Base64UrlUnpadded::encode_string(&key),
                        ..PublicKeyJwk::default()
                    })
                }
                _ => panic!("unsupported purpose"),
            }
        }
    }

    // HACK: generate a key pair
    #[allow(dead_code)]
    pub fn generate() -> Vec<u8> {
        // TODO: pass in public key
        let mut csprng = OsRng;
        let signing_key: SigningKey = SigningKey::generate(&mut csprng);
        let secret = Base64UrlUnpadded::encode_string(signing_key.as_bytes());
        println!("signing: {secret}");

        signing_key.verifying_key().to_bytes().to_vec()
    }
}
