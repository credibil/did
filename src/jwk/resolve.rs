//! Resolution of a DID into a public key using the `did:jwk` method.

use base64ct::{Base64UrlUnpadded, Encoding};

use crate::{Error, Method, MethodType, Resource, Url, VerificationMethodBuilder, VmKeyId};

/// Convert a `did:jwk` URL into a [`VerificationMethod`] object.
/// 
/// # Errors
/// If the URL is not a valid `did:jwk` URL, an error is returned. 
pub fn resolve(url: &Url) -> crate::Result<Resource> {
    if url.method != Method::Jwk {
        return Err(Error::InvalidDid(format!("DID is not a valid did:jwk: {url}")));
    }
    // For JWK, the method-specific identifier is the encoded key.
    let decoded = Base64UrlUnpadded::decode_vec(&url.id)
        .map_err(|e| Error::InvalidDid(format!("issue decoding key: {e}")))?;
    let jwk = serde_json::from_slice(&decoded)
        .map_err(|e| Error::InvalidDid(format!("issue deserializing key: {e}")))?;

    let vm = VerificationMethodBuilder::new(&jwk)
        .key_id(&url.resource_id(), VmKeyId::Index(String::new(), 0))?
        .method_type(&MethodType::JsonWebKey2020)?
        .build();
    
    Ok(Resource::VerificationMethod(vm))
}