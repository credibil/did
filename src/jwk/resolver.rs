//! # DID Key Resolver
//!
//! The `did:key` method is a DID method for static cryptographic keys. At its
//! core, it is based on expanding a cryptographic public key into a DID
//! Document.
//!
//! See:
//!
//! - <https://w3c-ccg.github.io/did-method-key>
//! - <https://w3c.github.io/did-resolution>

use std::sync::LazyLock;

use base64ct::{Base64UrlUnpadded, Encoding};
use regex::Regex;
use serde_json::json;

use super::DidJwk;
use crate::document::{CreateOptions, MethodType};
use crate::error::Error;
use crate::resolution::{ContentType, Metadata, Options, Resolved};
use crate::{DidOperator, DidResolver, KeyPurpose, PublicKeyJwk};

static DID_REGEX: LazyLock<Regex> =
    LazyLock::new(|| Regex::new("^did:jwk:(?<jwk>[A-Za-z0-9-=—]+)$").expect("should compile"));

struct Operator(MethodType);
impl DidOperator for Operator {
    fn verification(&self, purpose: KeyPurpose) -> Option<PublicKeyJwk> {
        match purpose {
            KeyPurpose::VerificationMethod => self.0.jwk().ok(),
            _ => panic!("unsupported purpose"),
        }
    }
}

impl DidJwk {
    pub fn resolve(did: &str, _: Option<Options>, _: impl DidResolver) -> crate::Result<Resolved> {
        // check DID is valid AND extract key
        let Some(caps) = DID_REGEX.captures(did) else {
            return Err(Error::InvalidDid("DID is not a valid did:jwk".into()));
        };
        // let enc = &caps["jwk"];

        let decoded = Base64UrlUnpadded::decode_vec(&caps["jwk"])
            .map_err(|e| Error::InvalidDid(format!("issue decoding key: {e}")))?;
        let jwk = serde_json::from_slice(&decoded)
            .map_err(|e| Error::InvalidDid(format!("issue deserializing key: {e}")))?;
        let op = Operator(MethodType::JsonWebKey { public_key_jwk: jwk });

        // per the spec, use the create operation to generate a DID document
        let options = CreateOptions {
            enable_encryption_key_derivation: true,
            ..CreateOptions::default()
        };

        let document = Self::create(op, options).map_err(|e| Error::InvalidDid(e.to_string()))?;

        Ok(Resolved {
            context: "https://w3id.org/did-resolution/v1".into(),
            metadata: Metadata {
                content_type: ContentType::DidLdJson,
                additional: Some(json!({
                    "pattern": "^did:jwk:[A-Za-z0-9-=—]+$",
                    "did": {
                        "didString": did,
                        "methodSpecificId": did[8..],
                        "method": "jwk"
                    }
                })),
                ..Metadata::default()
            },
            document: Some(document),
            ..Resolved::default()
        })
    }
}

#[cfg(test)]
mod test {
    use insta::assert_json_snapshot as assert_snapshot;

    use super::*;
    use crate::document::Document;

    const DID: &str = "did:jwk:eyJrdHkiOiJFQyIsImNydiI6InNlY3AyNTZrMSIsIngiOiJKSnpQaTRxeTJydktTVk85RjItMDVWV2VYMm9oc3dYN1NUbzg3TUdxcVB3IiwieSI6IkMxUnRGbnFXOWxOTEI1ejcycG9uMTIzZHh2MWtEcVUzUWw1QjhzMFdjXzQifQ";

    #[derive(Clone)]
    struct MockResolver;
    impl DidResolver for MockResolver {
        async fn resolve(&self, _url: &str) -> anyhow::Result<Document> {
            Ok(Document::default())
        }
    }

    #[tokio::test]
    async fn resolve() {
        let resolved = DidJwk::resolve(DID, None, MockResolver).expect("should resolve");
        assert_snapshot!("resolved", resolved);
    }
}
