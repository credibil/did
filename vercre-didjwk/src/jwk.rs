use vercre_ephemeral_keyring::{EphemeralKeyRing, KeyPair};

/// Registrar that implements the DID JWK method.
pub struct Registrar<K>
where
    K: KeyPair + Send + Sync,
{
    /// Key ring for managing keys and signing.
    pub(crate) keyring: EphemeralKeyRing<K>,
}

/// Configuration and internals.
impl<K> Registrar<K>
where
    K: KeyPair + Send + Sync,
{
    /// Create a new registrar.
    /// 
    /// # Arguments
    /// 
    /// * `keyring` - The keyring to use for generating a signing/verification key pair.
    /// 
    /// # Returns
    /// 
    /// * The registrar with the keyring configured.
    pub fn new(keyring: EphemeralKeyRing<K>) -> Self {
        Self { keyring }
    }
}
