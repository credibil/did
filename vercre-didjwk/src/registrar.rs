use base64ct::{Base64UrlUnpadded, Encoding};
use vercre_didcore::{
    error::Err, tracerr, Action, Context, DidDocument, KeyOperation, KeyPurpose, KeyRing, Patch,
    Registrar, Result, Service, VerificationMethod, VmWithPurpose, DID_CONTEXT,
};
use vercre_ephemeral_keyring::KeyPair;

use crate::jwk::Registrar as JwkRegistrar;

/// DID Registrar implementation for the JWK method.
#[allow(async_fn_in_trait)]
impl<K> Registrar for JwkRegistrar<K>
where
    K: KeyPair + Send + Sync,
{
    /// Create a DID document.
    ///
    /// For the JWK method, this will create a DID document with a single verification method. The
    /// services are not supported by this method and will return an error if specified.
    ///
    /// # Arguments
    ///
    /// * `services` - Must be None for this DID method.
    ///
    /// # Returns
    ///
    /// * The DID document. The ID of the document will be the DID.
    ///
    /// # Errors
    ///
    /// * If the services argument is not None.
    /// * If the keyring does not have a key for signing. This will happen if the Registrar's `new`
    /// function has not been called.
    /// * Other errors may be returned by the keyring if it is unable to generate a key.
    async fn create(&self, services: Option<&[Service]>) -> Result<DidDocument> {
        if services.is_some() {
            tracerr!(
                Err::NotSupported,
                "services are not supported by the JWK method"
            );
        }
        let signing_key = self.keyring.next_key(&KeyOperation::Sign).await?;
        let serialized = serde_json::to_vec(&signing_key)?;
        let encoded = Base64UrlUnpadded::encode_string(&serialized);
        let did = format!("did:{}:{}", Self::method(), encoded);

        let mut doc = DidDocument {
            context: vec![Context {
                url: Some(DID_CONTEXT.to_string()),
                ..Default::default()
            }],
            id: did.clone(),
            ..Default::default()
        };

        let mut vm = VmWithPurpose {
            verification_method: VerificationMethod {
                id: format!("{}#0", did.clone()),
                controller: did.clone(),
                type_: K::key_type().cryptosuite(),
                public_key_jwk: Some(signing_key.clone()),
                ..Default::default()
            },
            purposes: Some(vec![
                KeyPurpose::Authentication,
                KeyPurpose::AssertionMethod,
                KeyPurpose::CapabilityDelegation,
                KeyPurpose::CapabilityInvocation,
            ]),
        };
        if signing_key.use_ != Some("sig".to_string()) {
            if let Some(purposes) = vm.purposes.as_mut() {
                purposes.push(KeyPurpose::KeyAgreement);
            } else {
                vm.purposes = Some(vec![KeyPurpose::KeyAgreement]);
            }
        }
        let patch_key = Patch::builder(Action::AddPublicKeys).public_key(&vm)?.build()?;
        doc.apply_patches(&[patch_key]);

        Ok(doc)
    }

    /// The update operation is not supported for the JWK method.
    async fn update(&self, _: &DidDocument, _: &[Patch]) -> Result<DidDocument> {
        tracerr!(Err::NotSupported)
    }

    /// The deactivate operation is not supported for the JWK method.
    async fn deactivate(&self, _: &str) -> Result<()> {
        tracerr!(Err::NotSupported)
    }

    /// The recover operation is not supported for the JWK method.
    async fn recover(&self, _: &DidDocument) -> Result<()> {
        tracerr!(Err::NotSupported)
    }

    /// The method name for the JWK method.
    fn method() -> String {
        "jwk".to_owned()
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use vercre_ephemeral_keyring::{EphemeralKeyRing, Secp256k1KeyPair};

    #[tokio::test]
    async fn create() {
        let keyring = EphemeralKeyRing::<Secp256k1KeyPair>::new();
        let registrar = JwkRegistrar::new(keyring);
        let doc = registrar.create(None).await.unwrap();
        insta::with_settings!( {sort_maps => true}, {
            insta::assert_yaml_snapshot!(doc);
        });
    }
}
