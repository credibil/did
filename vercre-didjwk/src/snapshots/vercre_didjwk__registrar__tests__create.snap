---
source: vercre-didjwk/src/registrar.rs
assertion_line: 116
expression: doc
---
id: "did:jwk:eyJrdHkiOiJFQyIsImNydiI6InNlY3AyNTZrMSIsIngiOiJfYTN5WFh5WHpSOEVZZ3pubW1hQ09QeTJkUnY0UXM5NXd4RGdSX1pXUXJVIiwieSI6Imx4SzBLZy1QWE1tT2V6YnVpM1AtaEE3ZEdudmxHcFNHcGpjN0hUdlBkdUEifQ"
"@context": "https://www.w3.org/ns/did/v1"
verificationMethod:
  - id: "did:jwk:eyJrdHkiOiJFQyIsImNydiI6InNlY3AyNTZrMSIsIngiOiJfYTN5WFh5WHpSOEVZZ3pubW1hQ09QeTJkUnY0UXM5NXd4RGdSX1pXUXJVIiwieSI6Imx4SzBLZy1QWE1tT2V6YnVpM1AtaEE3ZEdudmxHcFNHcGpjN0hUdlBkdUEifQ#0"
    type: EcdsaSecp256k1VerificationKey2019
    controller: "did:jwk:eyJrdHkiOiJFQyIsImNydiI6InNlY3AyNTZrMSIsIngiOiJfYTN5WFh5WHpSOEVZZ3pubW1hQ09QeTJkUnY0UXM5NXd4RGdSX1pXUXJVIiwieSI6Imx4SzBLZy1QWE1tT2V6YnVpM1AtaEE3ZEdudmxHcFNHcGpjN0hUdlBkdUEifQ"
    publicKeyJwk:
      kty: EC
      crv: secp256k1
      x: _a3yXXyXzR8EYgznmmaCOPy2dRv4Qs95wxDgR_ZWQrU
      y: lxK0Kg-PXMmOezbui3P-hA7dGnvlGpSGpjc7HTvPduA
authentication:
  - "did:jwk:eyJrdHkiOiJFQyIsImNydiI6InNlY3AyNTZrMSIsIngiOiJfYTN5WFh5WHpSOEVZZ3pubW1hQ09QeTJkUnY0UXM5NXd4RGdSX1pXUXJVIiwieSI6Imx4SzBLZy1QWE1tT2V6YnVpM1AtaEE3ZEdudmxHcFNHcGpjN0hUdlBkdUEifQ#0"
assertionMethod:
  - "did:jwk:eyJrdHkiOiJFQyIsImNydiI6InNlY3AyNTZrMSIsIngiOiJfYTN5WFh5WHpSOEVZZ3pubW1hQ09QeTJkUnY0UXM5NXd4RGdSX1pXUXJVIiwieSI6Imx4SzBLZy1QWE1tT2V6YnVpM1AtaEE3ZEdudmxHcFNHcGpjN0hUdlBkdUEifQ#0"
keyAgreement:
  - "did:jwk:eyJrdHkiOiJFQyIsImNydiI6InNlY3AyNTZrMSIsIngiOiJfYTN5WFh5WHpSOEVZZ3pubW1hQ09QeTJkUnY0UXM5NXd4RGdSX1pXUXJVIiwieSI6Imx4SzBLZy1QWE1tT2V6YnVpM1AtaEE3ZEdudmxHcFNHcGpjN0hUdlBkdUEifQ#0"
capabilityInvocation:
  - "did:jwk:eyJrdHkiOiJFQyIsImNydiI6InNlY3AyNTZrMSIsIngiOiJfYTN5WFh5WHpSOEVZZ3pubW1hQ09QeTJkUnY0UXM5NXd4RGdSX1pXUXJVIiwieSI6Imx4SzBLZy1QWE1tT2V6YnVpM1AtaEE3ZEdudmxHcFNHcGpjN0hUdlBkdUEifQ#0"
capabilityDelegation:
  - "did:jwk:eyJrdHkiOiJFQyIsImNydiI6InNlY3AyNTZrMSIsIngiOiJfYTN5WFh5WHpSOEVZZ3pubW1hQ09QeTJkUnY0UXM5NXd4RGdSX1pXUXJVIiwieSI6Imx4SzBLZy1QWE1tT2V6YnVpM1AtaEE3ZEdudmxHcFNHcGpjN0hUdlBkdUEifQ#0"

